test/test_autograd.py

```diff
@@ -5761,6 +5763,134 @@ def test_saved_variable_packing_unpacking_did_not_save_original(self):
+        y.sum().backward()
+        self.assertEqual(2 * a, a.grad)
+
+    def test_default_saved_variable_hooks_did_not_save_original(self):
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+            a = torch.randn(5, requires_grad=True)
+            y = torch.exp(a)
+            self.assertEqual(y, y.grad_fn._saved_result)
+            y.sum().backward()
+            self.assertEqual(a.grad, y)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+    def test_setting_default_saved_variable_hooks_twice_should_fail(self):
+        try:
+            with self.assertRaisesRegex(RuntimeError, "Setting default hooks but they have already been set. "):
+                torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+                torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+    def test_saving_variable_to_disk(self):
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            def pack(x):
+                name = os.path.join(tmp_dir, str(uuid.uuid4()))
+                torch.save(x, name)
+                return name
+
+            def unpack(name):
+                return torch.load(name)
+
+            try:
+                torch.autograd.graph.set_saved_tensors_default_hooks(pack, unpack)
+                a = torch.ones(5, requires_grad=True)
+                y = a * a
+                self.assertEqual(a, y.grad_fn._saved_self)
+
+                y.sum().backward()
+                self.assertEqual(2 * a, a.grad)
+            finally:
+                torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+    def test_default_saved_variable_hooks_double_backward(self):
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: x, lambda x: x)
+            a = torch.randn(5, requires_grad=True)
+            y = a ** 3
+            s = torch.sum(y)
+            g, = torch.autograd.grad(s, (a, ), create_graph=True)
+            g.sum().backward()
+            self.assertEqual(6 * a, a.grad)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x)
+            a = torch.randn(5, requires_grad=True)
+            y = a ** 3
+            s = torch.sum(y)
+        finally:
+            torch.autograd.graph.reset_saved_tensors_default_hooks()
+        g, = torch.autograd.grad(s, (a, ), create_graph=True)
+        g.sum().backward()
+        self.assertEqual(6 * 2 * a, a.grad)
+
+        a = torch.randn(5, requires_grad=True)
+        y = a ** 3
+        s = torch.sum(y)
+        try:
+            torch.autograd.graph.set_saved_tensors_default_hooks(lambda x: 2 * x, lambda x: x)
```

Comment:
In summary: default hooks are applied to `SavedTensors` when the `grad_fn` that creates them has been created inside the context of those default hooks.

In that sense, the behavior of `torch.autograd.graph.set_saved_tensors_default_hooks` is very similar to `torch.set_grad_enabled`:

When grad is enabled (resp. when default hooks are set), a tensor that is created in that context will be given a `grad_fn` (resp. a `SavedVariable` that is created in that context will be given the default hooks and  packed according to the default pack hook).

In this case, `pow_backward` is `grad * (exp * self.pow(exp - 1))`.
`g.grad_fn` is a `MulBackward0` and  creates one `SavedVariable`: `g.grad_fn._saved_self`, which stores the value of `grad` (this is the first factor 2 from this test).
(`g.grad_fn._saved_other` is `None` because `grad` does not require gradients)
Then, `g.grad_fn.next_functions[1][0].next_functions[0][0]` is a `PowBackward0` which creates the `SavedVariable` : `g.grad_fn.next_functions[1][0].next_functions[0][0]._saved_self` which stores the value of `a` (second factor 2 of the test).

Note that `y.grad_fn._saved_self` also stores the value of `a` but is not affected by the hooks because `y` has been created outside of the context. 

The tensor `a` in itself does not have hooks. Only saved tensors do.